From 82c6d5e122e2b2ee90093116b384134997c7af95 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Mon, 3 Jul 2023 11:40:06 +0800
Subject: [PATCH 609/780] [Huawei] PSI: Introduce fine grained stall time
 collect for cgroup reclaim

Offering: HULK
HULK inclusion
category: feature
bugzilla: 188780

-------------------------------

PSI will tracking pressure stall for memory, cpu, io and irq. But, there
are differrnt pressure types which will cause memory pressure,
memory.pressure could not show the type of pressure effectively. The
same situation for cpu.pressure.
Introduce pressure.stat in psi, which will monitor specific reasons
for the memory.pressure and cpu.pressure, such as global/cgroup memory
reclaim, memory compact, cpu cfs bandwidth and so on. Therefore, userland
could make the right solution to reduce the pressure depends on the
specific pressure reasons.
This patch will introduce memory fine grained stall time collect for
cgroup reclaim.

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 include/linux/psi.h       |   6 ++
 include/linux/psi_types.h |  37 +++++++++
 include/linux/sched.h     |   4 +
 kernel/sched/psi.c        | 155 +++++++++++++++++++++++++++++++++++++-
 mm/memcontrol.c           |   3 +
 5 files changed, 201 insertions(+), 4 deletions(-)

diff --git a/include/linux/psi.h b/include/linux/psi.h
index a11c4a4ef469..41e01aa7508c 100644
--- a/include/linux/psi.h
+++ b/include/linux/psi.h
@@ -30,6 +30,12 @@ void psi_memstall_leave(unsigned long *flags);
 
 int psi_show(struct seq_file *s, struct psi_group *group, enum psi_res res);
 
+#ifdef CONFIG_PSI_FINE_GRAINED
+void psi_memstall_type_change(unsigned int type);
+#else
+static inline void psi_memstall_type_change(unsigned int type) {}
+#endif
+
 #ifdef CONFIG_CGROUPS
 int psi_cgroup_alloc(struct cgroup *cgrp);
 void psi_cgroup_free(struct cgroup *cgrp);
diff --git a/include/linux/psi_types.h b/include/linux/psi_types.h
index ec8ba7428c50..6abe782b7acd 100644
--- a/include/linux/psi_types.h
+++ b/include/linux/psi_types.h
@@ -229,11 +229,48 @@ struct psi_group {
 };
 
 #ifdef CONFIG_PSI_FINE_GRAINED
+
+enum psi_stat_states {
+	PSI_MEMCG_RECLAIM_SOME,
+	PSI_MEMCG_RECLAIM_FULL,
+	NR_PSI_STAT_STATES,
+};
+
+enum psi_stat_task_count {
+	NR_MEMCG_RECLAIM,
+	NR_MEMCG_RECLAIM_RUNNING,
+	NR_PSI_STAT_TASK_COUNTS,
+};
+
+/*
+ * one type should have two task stats: regular running and memstall
+ * threads. The reason is the same as NR_MEMSTALL_RUNNING.
+ * Because of the psi_memstall_type is start with 1, the correspondence
+ * between psi_memstall_type and psi_stat_task_count should be as below:
+ *
+ * memstall : psi_memstall_type * 2 - 2;
+ * running  : psi_memstall_type * 2 - 1;
+ */
+enum psi_memstall_type {
+	PSI_MEMCG_RECLAIM = 1,
+};
+
+struct psi_group_stat_cpu {
+	u32 state_mask;
+	u32 times[NR_PSI_STAT_STATES];
+	u32 psi_delta;
+	unsigned int tasks[NR_PSI_STAT_TASK_COUNTS];
+};
+
 struct psi_group_ext {
 	struct psi_group psi;
+	struct psi_group_stat_cpu __percpu *pcpu;
 };
 #else
+
 struct psi_group_ext { };
+#define PSI_MEMCG_RECLAIM		0
+
 #endif /* CONFIG_PSI_FINE_GRAINED */
 
 #else /* CONFIG_PSI */
diff --git a/include/linux/sched.h b/include/linux/sched.h
index ae8dfdae5143..942ac2c423b4 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1427,7 +1427,11 @@ struct task_struct {
 	KABI_RESERVE(2)
 #endif
 	KABI_USE(3, void *pf_io_worker)
+#ifdef CONFIG_PSI_FINE_GRAINED
+	KABI_USE(4, int memstall_type)
+#else
 	KABI_RESERVE(4)
+#endif
 	KABI_RESERVE(5)
 	KABI_RESERVE(6)
 	KABI_RESERVE(7)
diff --git a/kernel/sched/psi.c b/kernel/sched/psi.c
index 5b6188e21a71..ea4ba94eb788 100644
--- a/kernel/sched/psi.c
+++ b/kernel/sched/psi.c
@@ -194,7 +194,10 @@ struct psi_group psi_system = {
 
 #ifdef CONFIG_PSI_FINE_GRAINED
 /* System-level fine grained pressure and stall tracking */
-struct psi_group_ext psi_stat_system = { };
+static DEFINE_PER_CPU(struct psi_group_stat_cpu, system_stat_group_pcpu);
+struct psi_group_ext psi_stat_system = {
+	.pcpu = &system_stat_group_pcpu,
+};
 
 struct psi_group_ext *to_psi_group_ext(struct psi_group *psi)
 {
@@ -334,6 +337,120 @@ static void calc_avgs(unsigned long avg[3], int missed_periods,
 	avg[2] = calc_load(avg[2], EXP_300s, pct);
 }
 
+#ifdef CONFIG_PSI_FINE_GRAINED
+
+void psi_memstall_type_change(unsigned int type)
+{
+	struct rq_flags rf;
+	struct rq *rq;
+
+	if (static_branch_likely(&psi_disabled))
+		return;
+
+	rq = this_rq_lock_irq(&rf);
+	current->memstall_type = type;
+	rq_unlock_irq(rq, &rf);
+}
+
+static void record_stat_times(struct psi_group_ext *psi_ext, int cpu)
+{
+	struct psi_group_stat_cpu *ext_grpc = per_cpu_ptr(psi_ext->pcpu, cpu);
+
+	u32 delta = ext_grpc->psi_delta;
+
+	if (ext_grpc->state_mask & (1 << PSI_MEMCG_RECLAIM_SOME)) {
+		ext_grpc->times[PSI_MEMCG_RECLAIM_SOME] += delta;
+		if (ext_grpc->state_mask & (1 << PSI_MEMCG_RECLAIM_FULL))
+			ext_grpc->times[PSI_MEMCG_RECLAIM_FULL] += delta;
+	}
+}
+
+static bool test_fine_grained_stat(unsigned int *stat_tasks,
+				   unsigned int nr_running,
+				   enum psi_stat_states state)
+{
+	switch (state) {
+	case PSI_MEMCG_RECLAIM_SOME:
+		return unlikely(stat_tasks[NR_MEMCG_RECLAIM]);
+	case PSI_MEMCG_RECLAIM_FULL:
+		return unlikely(stat_tasks[NR_MEMCG_RECLAIM] &&
+		       nr_running == stat_tasks[NR_MEMCG_RECLAIM_RUNNING]);
+	default:
+		return false;
+	}
+}
+
+static void psi_group_stat_change(struct psi_group *group, int cpu,
+				  int clear, int set)
+{
+	int t;
+	u32 state_mask = 0;
+	enum psi_stat_states s;
+	struct psi_group_ext *psi_ext = to_psi_group_ext(group);
+	struct psi_group_cpu *groupc = per_cpu_ptr(group->pcpu, cpu);
+	struct psi_group_stat_cpu *ext_groupc = per_cpu_ptr(psi_ext->pcpu, cpu);
+
+	write_seqcount_begin(&groupc->seq);
+	record_stat_times(psi_ext, cpu);
+
+	for (t = 0; clear; clear &= ~(1 << t), t++)
+		if (clear & (1 << t))
+			ext_groupc->tasks[t]--;
+	for (t = 0; set; set &= ~(1 << t), t++)
+		if (set & (1 << t))
+			ext_groupc->tasks[t]++;
+	for (s = 0; s < NR_PSI_STAT_STATES; s++)
+		if (test_fine_grained_stat(ext_groupc->tasks,
+					   groupc->tasks[NR_RUNNING], s))
+			state_mask |= (1 << s);
+	if (unlikely(groupc->state_mask & PSI_ONCPU) &&
+		     cpu_curr(cpu)->memstall_type)
+		state_mask |= (1 << (cpu_curr(cpu)->memstall_type * 2 - 1));
+
+	ext_groupc->state_mask = state_mask;
+	write_seqcount_end(&groupc->seq);
+}
+
+static void update_psi_stat_delta(struct psi_group *group, int cpu, u64 now)
+{
+	struct psi_group_ext *psi_ext = to_psi_group_ext(group);
+	struct psi_group_stat_cpu *ext_groupc = per_cpu_ptr(psi_ext->pcpu, cpu);
+	struct psi_group_cpu *groupc = per_cpu_ptr(group->pcpu, cpu);
+
+	ext_groupc->psi_delta = now - groupc->state_start;
+}
+
+static void psi_stat_flags_change(struct task_struct *task, int *stat_set,
+				  int *stat_clear, int set, int clear)
+{
+	if (!task->memstall_type)
+		return;
+
+	if (clear) {
+		if (clear & TSK_MEMSTALL)
+			*stat_clear |= 1 << (2 * task->memstall_type - 2);
+		if (clear & TSK_MEMSTALL_RUNNING)
+			*stat_clear |= 1 << (2 * task->memstall_type - 1);
+	}
+	if (set) {
+		if (set & TSK_MEMSTALL)
+			*stat_set |= 1 << (2 * task->memstall_type - 2);
+		if (set & TSK_MEMSTALL_RUNNING)
+			*stat_set |= 1 << (2 * task->memstall_type - 1);
+	}
+}
+
+#else
+static inline void psi_group_stat_change(struct psi_group *group, int cpu,
+					 int clear, int set) {}
+static inline void update_psi_stat_delta(struct psi_group *group, int cpu,
+					 u64 now) {}
+static inline void psi_stat_flags_change(struct task_struct *task,
+					 int *stat_set, int *stat_clear,
+					 int set, int clear) {}
+static inline void record_stat_times(struct psi_group_ext *psi_ext, int cpu) {}
+#endif
+
 static void collect_percpu_times(struct psi_group *group,
 				 enum psi_aggregators aggregator,
 				 u32 *pchanged_states)
@@ -853,16 +970,22 @@ void psi_task_change(struct task_struct *task, int clear, int set)
 	struct psi_group *group;
 	void *iter = NULL;
 	u64 now;
+	int stat_set = 0;
+	int stat_clear = 0;
 
 	if (!task->pid)
 		return;
 
 	psi_flags_change(task, clear, set);
+	psi_stat_flags_change(task, &stat_set, &stat_clear, set, clear);
 
 	now = cpu_clock(cpu);
 
-	while ((group = iterate_groups(task, &iter)))
+	while ((group = iterate_groups(task, &iter))) {
+		update_psi_stat_delta(group, cpu, now);
 		psi_group_change(group, cpu, clear, set, now, true);
+		psi_group_stat_change(group, cpu, stat_clear, stat_set);
+	}
 }
 
 void psi_task_switch(struct task_struct *prev, struct task_struct *next,
@@ -889,13 +1012,17 @@ void psi_task_switch(struct task_struct *prev, struct task_struct *next,
 				break;
 			}
 
+			update_psi_stat_delta(group, cpu, now);
 			psi_group_change(group, cpu, 0, TSK_ONCPU, now, true);
+			psi_group_stat_change(group, cpu, 0, 0);
 		}
 	}
 
 	if (prev->pid) {
 		int clear = TSK_ONCPU, set = 0;
 		bool wake_clock = true;
+		int stat_set = 0;
+		int stat_clear = 0;
 
 		/*
 		 * When we're going to sleep, psi_dequeue() lets us
@@ -922,10 +1049,14 @@ void psi_task_switch(struct task_struct *prev, struct task_struct *next,
 		}
 
 		psi_flags_change(prev, clear, set);
+		psi_stat_flags_change(prev, &stat_set, &stat_clear, set, clear);
 
 		iter = NULL;
-		while ((group = iterate_groups(prev, &iter)) && group != common)
+		while ((group = iterate_groups(prev, &iter)) && group != common) {
+			update_psi_stat_delta(group, cpu, now);
 			psi_group_change(group, cpu, clear, set, now, wake_clock);
+			psi_group_stat_change(group, cpu, stat_clear, stat_set);
+		}
 
 		/*
 		 * TSK_ONCPU is handled up to the common ancestor. If there are
@@ -935,8 +1066,12 @@ void psi_task_switch(struct task_struct *prev, struct task_struct *next,
 		 */
 		if ((prev->psi_flags ^ next->psi_flags) & ~TSK_ONCPU) {
 			clear &= ~TSK_ONCPU;
-			for (; group; group = iterate_groups(prev, &iter))
+			for (; group; group = iterate_groups(prev, &iter)) {
+				update_psi_stat_delta(group, cpu, now);
 				psi_group_change(group, cpu, clear, set, now, wake_clock);
+				psi_group_stat_change(group, cpu, stat_clear,
+						      stat_set);
+			}
 		}
 	}
 }
@@ -963,6 +1098,8 @@ void psi_account_irqtime(struct task_struct *task, u32 delta)
 
 		write_seqcount_begin(&groupc->seq);
 
+		update_psi_stat_delta(group, cpu, now);
+		record_stat_times(to_psi_group_ext(group), cpu);
 		record_times(groupc, now);
 		groupc->times[PSI_IRQ_FULL] += delta;
 
@@ -1035,6 +1172,9 @@ void psi_memstall_leave(unsigned long *flags)
 
 	current->in_memstall = 0;
 	psi_task_change(current, TSK_MEMSTALL | TSK_MEMSTALL_RUNNING, 0);
+#ifdef CONFIG_PSI_FINE_GRAINED
+	current->memstall_type = 0;
+#endif
 
 	rq_unlock_irq(rq, &rf);
 }
@@ -1053,6 +1193,11 @@ int psi_cgroup_alloc(struct cgroup *cgroup)
 	psi_ext = kzalloc(sizeof(struct psi_group_ext), GFP_KERNEL);
 	if (!psi_ext)
 		return -ENOMEM;
+	psi_ext->pcpu = alloc_percpu(struct psi_group_stat_cpu);
+	if (!psi_ext->pcpu) {
+		kfree(psi_ext);
+		return -ENOMEM;
+	}
 	cgroup->psi = &psi_ext->psi;
 #else
 	cgroup->psi = kzalloc(sizeof(struct psi_group), GFP_KERNEL);
@@ -1063,6 +1208,7 @@ int psi_cgroup_alloc(struct cgroup *cgroup)
 	cgroup->psi->pcpu = alloc_percpu(struct psi_group_cpu);
 	if (!cgroup->psi->pcpu) {
 #ifdef CONFIG_PSI_FINE_GRAINED
+		free_percpu(psi_ext->pcpu);
 		kfree(psi_ext);
 #else
 		kfree(cgroup->psi);
@@ -1083,6 +1229,7 @@ void psi_cgroup_free(struct cgroup *cgroup)
 	/* All triggers must be removed by now */
 	WARN_ONCE(cgroup->psi->poll_states, "psi: trigger leak\n");
 #ifdef CONFIG_PSI_FINE_GRAINED
+	free_percpu(to_psi_group_ext(cgroup->psi)->pcpu);
 	kfree(to_psi_group_ext(cgroup->psi));
 #else
 	kfree(cgroup->psi);
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 7e83c47cc1ea..870af5412663 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -2360,6 +2360,7 @@ static unsigned long reclaim_high(struct mem_cgroup *memcg,
 
 		memcg_memory_event(memcg, MEMCG_HIGH);
 
+		psi_memstall_type_change(PSI_MEMCG_RECLAIM);
 		psi_memstall_enter(&pflags);
 		nr_reclaimed += try_to_free_mem_cgroup_pages(memcg, nr_pages,
 							     gfp_mask, true);
@@ -2630,6 +2631,7 @@ void mem_cgroup_handle_over_high(void)
 	 * schedule_timeout_killable sets TASK_KILLABLE). This means we don't
 	 * need to account for any ill-begotten jiffies to pay them off later.
 	 */
+	psi_memstall_type_change(PSI_MEMCG_RECLAIM);
 	psi_memstall_enter(&pflags);
 	schedule_timeout_killable(penalty_jiffies);
 	psi_memstall_leave(&pflags);
@@ -2701,6 +2703,7 @@ static int try_charge(struct mem_cgroup *memcg, gfp_t gfp_mask,
 
 	memcg_memory_event(mem_over_limit, MEMCG_MAX);
 
+	psi_memstall_type_change(PSI_MEMCG_RECLAIM);
 	psi_memstall_enter(&pflags);
 	nr_reclaimed = try_to_free_mem_cgroup_pages(mem_over_limit, nr_pages,
 						    gfp_mask, may_swap);
-- 
2.17.1

